File: examples/zig-map/map_state.zig
const std = @import("std");
const maplibre = @import("maplibre_native_ffi");

const diagnostics = @import("diagnostics.zig");
const types = @import("types.zig");

pub const MapState = struct {
    allocator: std.mem.Allocator,
    diagnostic_store: *maplibre.DiagnosticStore,
    runtime: maplibre.RuntimeHandle,
    map: maplibre.MapHandle,

    pub fn init(allocator: std.mem.Allocator, viewport: types.Viewport) !MapState {
        const diagnostic_store = try allocator.create(maplibre.DiagnosticStore);
        diagnostic_store.* = maplibre.DiagnosticStore.init(allocator);
        errdefer {
            diagnostic_store.deinit();
            allocator.destroy(diagnostic_store);
        }

        var runtime = maplibre.RuntimeHandle.create(allocator, .{ .cache_path = ":memory:" }, diagnostic_store) catch |err| {
            diagnostics.logError("runtime create failed", err, diagnostic_store);
            return types.AppError.RuntimeCreateFailed;
        };
        errdefer if (runtime.close()) |future| {
            var teardown = future;
            _ = teardown.wait(null) catch {};
            teardown.deinit();
        } else |_| {};

        // Selecting the event mask at creation puts it ahead of the style
        // load, so the map queues render updates from the first tile response.
        // The render loop re-arms from the frame result's repaint flag, so the
        // map only has to report updates that arrive between frames.
        var map_future = maplibre.MapHandle.create(&runtime, .{
            .width = viewport.logical_width,
            .height = viewport.logical_height,
            .scale_factor = viewport.scale_factor,
            .mode = .continuous,
            .event_mask = .{ .map_render_update_available = true },
        }) catch |err| {
            diagnostics.logError("map create failed", err, diagnostic_store);
            return types.AppError.MapCreateFailed;
        };
        defer map_future.deinit();
        var map = map_future.wait(diagnostic_store) catch |err| {
            diagnostics.logError("map create failed", err, diagnostic_store);
            return types.AppError.MapCreateFailed;
        };
        errdefer if (map.close()) |future| {
            var teardown = future;
            teardown.deinit();
        } else |_| {};

        try loadStyle(allocator, &map, diagnostic_store);
        try setCamera(&map, diagnostic_store);
        return .{
            .allocator = allocator,
            .diagnostic_store = diagnostic_store,
            .runtime = runtime,
            .map = map,
        };
    }

    pub fn deinit(self: *MapState) void {
        // Awaiting both release completions keeps process exit ordered after
        // native teardown.
        if (self.map.close()) |future| {
            var teardown = future;
            _ = teardown.wait(null) catch {};
            teardown.deinit();
        } else |_| {}
        if (self.runtime.close()) |future| {
            var teardown = future;
            _ = teardown.wait(null) catch {};
            teardown.deinit();
        } else |_| {}
        self.diagnostic_store.deinit();
        self.allocator.destroy(self.diagnostic_store);
    }

    pub fn setGesture(self: *MapState, phase: maplibre.GesturePhase) !void {
        try self.updateCamera(.{ .gesture_phase = phase });
    }

    pub fn moveBy(self: *MapState, dx: f64, dy: f64) !void {
        try self.cameraMutation(self.map.applyCameraDelta(.{ .offset = .{ .x = dx, .y = dy } }));
    }

    pub fn moveByAnimated(self: *MapState, dx: f64, dy: f64, duration_ms: f64) !void {
        try self.cameraMutation(self.map.applyCameraDelta(.{
            .offset = .{ .x = dx, .y = dy },
            .animation = .{ .duration_ms = duration_ms },
        }));
    }

    pub fn scaleBy(self: *MapState, scale: f64, anchor: maplibre.ScreenPoint) !void {
        try self.cameraMutation(self.map.applyCameraDelta(.{ .kind = .scale, .amount = scale, .anchor = anchor }));
    }

    pub fn scaleByAnimated(self: *MapState, scale: f64, anchor: maplibre.ScreenPoint, duration_ms: f64) !void {
        try self.cameraMutation(self.map.applyCameraDelta(.{
            .kind = .scale,
            .amount = scale,
            .anchor = anchor,
            .animation = .{ .duration_ms = duration_ms },
        }));
    }

    pub fn pitchBy(self: *MapState, delta: f64) !void {
        try self.cameraMutation(self.map.applyCameraDelta(.{ .kind = .pitch, .amount = delta }));
    }

    pub fn adjustBearing(self: *MapState, delta: f64) !void {
        try self.cameraMutation(self.map.applyCameraDelta(.{ .kind = .bearing, .amount = delta }));
    }

    pub fn adjustBearingAnimated(self: *MapState, delta: f64, duration_ms: f64) !void {
        try self.cameraMutation(self.map.applyCameraDelta(.{
            .kind = .bearing,
            .amount = delta,
            .animation = .{ .duration_ms = duration_ms },
        }));
    }

    pub fn adjustPitchAnimated(self: *MapState, delta: f64, duration_ms: f64) !void {
        try self.cameraMutation(self.map.applyCameraDelta(.{
            .kind = .pitch,
            .amount = delta,
            .animation = .{ .duration_ms = duration_ms },
        }));
    }

    pub fn resetOrientation(self: *MapState, duration_ms: f64) !void {
        try self.updateCamera(.{
            .mode = .ease,
            .camera = .{ .bearing = 0, .pitch = 0 },
            .animation = .{ .duration_ms = duration_ms },
        });
    }

    fn updateCamera(self: *MapState, update: maplibre.CameraUpdate) !void {
        try self.cameraMutation(self.map.updateCamera(update));
    }

    /// Ends any running camera transition, so a starting gesture takes over
    /// from it rather than fighting it.
    pub fn cancelTransitions(self: *MapState) !void {
        try self.cameraMutation(self.map.cancelTransitions());
    }

    /// Drains runtime events, reporting whether the map requested another
    /// frame.
    pub fn drainEvents(self: *MapState) !bool {
        const map_id = try self.map.id();
        var batch = try self.runtime.drainEvents(self.allocator);
        defer batch.deinit();
        for (0..batch.len()) |index| {
            const event = try batch.at(index);
            if (event.source_type != .map or event.source_id == null or
                !std.meta.eql(event.source_id.?, map_id)) continue;
            if (event.event_type == .map_render_update_available) return true;
        }
        return false;
    }

    fn cameraMutation(self: *MapState, result: anytype) !void {
        var completion = result catch |err| {
            diagnostics.logError("camera update failed", err, self.diagnostic_store);
            return types.AppError.CameraUpdateFailed;
        };
        completion.deinit();
    }
};

fn loadStyle(
    allocator: std.mem.Allocator,
    map: *maplibre.MapHandle,
    diagnostic_store: *const maplibre.DiagnosticStore,
) !void {
    var completion = map.setStyleUrl(allocator, "https://tiles.openfreemap.org/styles/bright") catch |err| {
        diagnostics.logError("style load failed", err, diagnostic_store);
        return types.AppError.StyleLoadFailed;
    };
    completion.deinit();
}

fn setCamera(
    map: *maplibre.MapHandle,
    diagnostic_store: *const maplibre.DiagnosticStore,
) !void {
    var completion = map.updateCamera(.{ .camera = .{
        .center = .{ .latitude = 37.7749, .longitude = -122.4194 },
        .zoom = 13.0,
        .bearing = 12.0,
        .pitch = 30.0,
    } }) catch |err| {
        diagnostics.logError("camera jump failed", err, diagnostic_store);
        return types.AppError.CameraJumpFailed;
    };
    completion.deinit();
}

File: examples/go-map/map_state.go
package main

import (
	"errors"
	"fmt"
	"math"

	maplibre "github.com/maplibre/maplibre-native-ffi/bindings/go"
)

type runtimeMapState struct {
	runtime *maplibre.RuntimeHandle
	mapRef  *maplibre.MapHandle
	mapID   maplibre.MapID
}

func newRuntimeMapState(v viewport) (*runtimeMapState, error) {
	runtimeHandle, err := maplibre.NewRuntimeWithOptions(maplibre.NewRuntimeOptions("", ":memory:"))
	if err != nil {
		return nil, fmt.Errorf("runtime create failed: %w", err)
	}
	state := &runtimeMapState{runtime: runtimeHandle}
	mapOptions := maplibre.NewMapOptions(v.logicalWidth, v.logicalHeight, v.scaleFactor)
	mapOptions.EventMask = maplibre.RuntimeEventMaskMapRenderUpdateAvailable |
		maplibre.RuntimeEventMaskMapRenderFrameFinished
	mapHandle, err := runtimeHandle.NewMapWithOptions(mapOptions)
	if err != nil {
		_ = state.Close()
		return nil, fmt.Errorf("map create failed: %w", err)
	}
	state.mapRef = mapHandle
	state.mapID, err = mapHandle.ID()
	if err != nil {
		_ = state.Close()
		return nil, fmt.Errorf("map identity read failed: %w", err)
	}
	_, err = mapHandle.SetStyleURL("https://tiles.openfreemap.org/styles/bright")
	if err != nil {
		_ = state.Close()
		return nil, fmt.Errorf("style load failed: %w", err)
	}
	initialCamera := maplibre.CameraOptions{}.
		WithCenter(maplibre.LatLng{Latitude: 37.7749, Longitude: -122.4194}).
		WithZoom(13).
		WithBearing(12).
		WithPitch(30)
	_, err = mapHandle.JumpTo(initialCamera)
	if err != nil {
		_ = state.Close()
		return nil, fmt.Errorf("camera jump failed: %w", err)
	}
	barrier, err := runtimeHandle.Barrier()
	if err != nil {
		_ = state.Close()
		return nil, fmt.Errorf("initial barrier failed: %w", err)
	}
	if completed, err := barrier.Wait(-1); err != nil || !completed {
		barrier.Release()
		_ = state.Close()
		return nil, fmt.Errorf("initial barrier wait failed: completed=%v: %w", completed, err)
	}
	if err := barrier.Discard(); err != nil {
		barrier.Release()
		_ = state.Close()
		return nil, fmt.Errorf("initial barrier result failed: %w", err)
	}
	barrier.Release()
	_, err = mapHandle.RequestRepaint()
	if err != nil {
		_ = state.Close()
		return nil, fmt.Errorf("initial repaint request failed: %w", err)
	}
	return state, nil
}

func (state *runtimeMapState) Close() error {
	var result error
	if state.mapRef != nil {
		result = errors.Join(result, state.mapRef.Close())
		state.mapRef = nil
	}
	if state.runtime != nil {
		result = errors.Join(result, state.runtime.Close())
		state.runtime = nil
	}
	return result
}

func (state *runtimeMapState) cancelTransitions() error {
	return state.updateCamera(maplibre.CameraUpdate{GesturePhase: maplibre.GesturePhaseCancel})
}

func (state *runtimeMapState) setGestureInProgress(inProgress bool) error {
	phase := maplibre.GesturePhaseEnd
	if inProgress {
		phase = maplibre.GesturePhaseBegin
	}
	return state.updateCamera(maplibre.CameraUpdate{GesturePhase: phase})
}

func (state *runtimeMapState) moveBy(dx, dy float64, durationMS *float64) error {
	camera, err := state.orderedCamera()
	if err != nil {
		return fmt.Errorf("ordered camera read failed: %w", err)
	}
	center := maplibre.LatLng{}
	if camera.Center != nil {
		center = *camera.Center
	}
	zoom := 0.0
	if camera.Zoom != nil {
		zoom = *camera.Zoom
	}
	degreesPerPixel := 360 / (256 * math.Exp2(zoom))
	center.Longitude -= dx * degreesPerPixel
	center.Latitude += dy * degreesPerPixel
	return state.updateCamera(cameraUpdate(maplibre.CameraOptions{}.WithCenter(center), durationMS))
}

func (state *runtimeMapState) scaleBy(scale float64, anchor maplibre.ScreenPoint, durationMS *float64) error {
	camera, err := state.orderedCamera()
	if err != nil {
		return fmt.Errorf("ordered camera read failed: %w", err)
	}
	zoom := 0.0
	if camera.Zoom != nil {
		zoom = *camera.Zoom
	}
	options := maplibre.CameraOptions{}.WithZoom(zoom + math.Log2(scale))
	options.Anchor = &anchor
	return state.updateCamera(cameraUpdate(options, durationMS))
}

func (state *runtimeMapState) adjustPitch(delta float64, durationMS *float64) error {
	camera, err := state.orderedCamera()
	if err != nil {
		return fmt.Errorf("ordered camera read failed: %w", err)
	}
	pitch := delta
	if camera.Pitch != nil {
		pitch += *camera.Pitch
	}
	return state.updateCamera(cameraUpdate(maplibre.CameraOptions{}.WithPitch(clamp(pitch, 0, 60)), durationMS))
}

func (state *runtimeMapState) adjustBearing(delta float64, durationMS *float64) error {
	camera, err := state.orderedCamera()
	if err != nil {
		return fmt.Errorf("ordered camera read failed: %w", err)
	}
	bearing := delta
	if camera.Bearing != nil {
		bearing += *camera.Bearing
	}
	return state.updateCamera(cameraUpdate(maplibre.CameraOptions{}.WithBearing(bearing), durationMS))
}

func (state *runtimeMapState) resetOrientation(durationMS float64) error {
	return state.updateCamera(cameraUpdate(maplibre.CameraOptions{}.WithBearing(0).WithPitch(0), &durationMS))
}

func (state *runtimeMapState) updateCamera(update maplibre.CameraUpdate) error {
	if _, err := state.mapRef.UpdateCamera(update); err != nil {
		return fmt.Errorf("camera update failed: %w", err)
	}
	return nil
}

func cameraUpdate(options maplibre.CameraOptions, durationMS *float64) maplibre.CameraUpdate {
	update := maplibre.CameraUpdate{Camera: options}
	if durationMS != nil {
		animation := maplibre.AnimationOptions{}.WithDurationMS(*durationMS)
		update.Mode = maplibre.CameraUpdateModeEase
		update.Animation = &animation
	}
	return update
}

func (state *runtimeMapState) orderedCamera() (maplibre.CameraOptions, error) {
	operation, err := state.mapRef.QueryCamera()
	if err != nil {
		return maplibre.CameraOptions{}, err
	}
	defer operation.Release()
	if completed, err := operation.Wait(-1); err != nil {
		return maplibre.CameraOptions{}, err
	} else if !completed {
		return maplibre.CameraOptions{}, errors.New("camera query returned before completion")
	}
	result, err := operation.Take()
	return result.Camera, err
}

func drainEvents(runtimeHandle *maplibre.RuntimeHandle, mapID maplibre.MapID) (bool, error) {
	renderRequested := false
	batch, err := runtimeHandle.DrainEvents()
	if err != nil {
		return false, fmt.Errorf("runtime event drain failed: %w", err)
	}
	for _, event := range batch.Events {
		if event.Source.Type != maplibre.RuntimeEventSourceMap || event.Source.MapID != mapID {
			continue
		}
		switch event.Type {
		case maplibre.RuntimeEventMapRenderUpdateAvailable:
			renderRequested = true
		case maplibre.RuntimeEventMapRenderFrameFinished:
			payload, ok := event.Payload.(maplibre.RuntimeEventRenderFramePayload)
			if ok && payload.NeedsRepaint {
				renderRequested = true
			}
		}
	}
	return renderRequested, nil
}

// renderMapState owns the render target on the SDL render loop thread.
type renderMapState struct {
	target renderTarget
}

func newRenderMapState(graphics *openGLContext, mapRef *maplibre.MapHandle, v viewport, mode renderTargetMode) (*renderMapState, error) {
	target, err := newOpenGLRenderTarget(graphics, v, mode, mapRef)
	if err != nil {
		return nil, err
	}
	return &renderMapState{target: target}, nil
}

func (state *renderMapState) closeTarget() error {
	if state.target == nil {
		return nil
	}
	err := state.target.Close()
	state.target = nil
	return err
}

func (state *renderMapState) resize(v viewport) error {
	if state.target == nil {
		return errors.New("render target is not attached")
	}
	return state.target.Resize(v)
}

func (state *renderMapState) finishFrame() error {
	if state.target == nil {
		return nil
	}
	return state.target.FinishFrame()
}

func (state *renderMapState) driveFrame() (bool, error) {
	if state.target == nil {
		return false, nil
	}
	return state.target.DriveFrame()
}
